### PR TITLE
Adds follow (pubsub) feature

### LINKED ISSUE #739: Follow Patrons & activity feeds
Related to: #857, and #1964

See: #8607

## Ideas
And possibly receive email updates when users create lists or mark books on their Reading Logs.

### PR DESCRIPTION
**Pull request by mekarpeles**
* Followup to #8597, closes #1966 (previous PR) and implements #739

closes #739

- database schema for follows
- adds core model for follows
- elementary mybooks /feed of followers
- ability to follow/unfollow
- json API to fetch patron's follows

### Screen shots

**My Feed** (private)
Screenshot 2024-02-20 at 9 48 08 AM

**My Followers** (public)
Screenshot 2024-02-20 at 9 48 31 AM

**Following** (public)
Screenshot 2024-02-20 at 9 48 59 AM

### Technical

**Schema already on production db**

### Product Notes

- [ ] Add follows to /stats dashboard
- [x] Add Patrons I Follow page to my books.
  - There are still valuable pieces from #1966 which we should consider implementing, like the [people I'm following page](https://github.com/internetarchive/openlibrary/pull/1966/files#diff-640b35cc042bf968b008b029cc8c77593cb5b668d22fa787cef684d31b4937f9) even if this is private
- [ ] Leverage follows to integrate book recommendations into Books Page
using with the following code:

```
+++ b/openlibrary/templates/type/edition/view.html
@@ -338,6 +338,13 @@ $ worldcat_links = macros.WorldcatLink(isbn=isbn_13 or isbn_10, oclc_numbers=ocl
       </div>
     </div>
 
+    <div>
+      <ul>
+	$for p in page.get_patrons_who_also_read():
+	  <a href="/people/$(p.key.split('/')[2])/books/already-read">$p.key.split('/')[2]</a></li>
+      </ul>
+    </div>
```

### CODE REVIEW

**Comment by mekarpeles** [comment]
To leverage follows, on the Books page, start with the following code:

```
+++ b/openlibrary/templates/type/edition/view.html
@@ -338,6 +338,13 @@ $ worldcat_links = macros.WorldcatLink(isbn=isbn_13 or isbn_10, oclc_numbers=ocl
       </div>
     </div>
 
+    <div>
+      <ul>
+	$for p in page.get_patrons_who_also_read():
+	  <a href="/people/$(p.key.split('/')[2])/books/already-read">$p.key.split('/')[2]</a></li>
+      </ul>
+    </div>
```

**Comment by mekarpeles** [comment]
BLOCKER: Bug,
http://ol-dev1.us.archive.org:1338/people/seabelis/followers seabelis following themself?

File: openlibrary/plugins/upstream/account.py [hunk 1]
```diff
@@ -1055,6 +1056,21 @@ def format_rating(rating: Mapping) -> dict:
...
+
+    def GET(self, username, key=""):
+        mb = MyBooksTemplate(username, 'following')
+        follows = (
+            PubSub.get_subscribers(username)
+            if key == 'followers'
+            else PubSub.get_subscriptions(username)
+        )
+        manage = key == 'following'
+        template = render['account/follows'](mb.me, follows, manage=manage)
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/account.py:1070]:
```suggestion
        template = render['account/follows'](mb.user, follows, manage=manage)
```

**Review by jimchamp — CHANGES_REQUESTED** [review]
Thanks, @mekarpeles!  A few things that I've noticed:

1. The "My Feeds" page doesn't render for patrons who are not following anybody.  The issue is with the `get_feed` function.  I've added some notes about this in the review, but couldn't come up with a solution that didn't cause other bugs...
2. I don't think that `PubSub` should extend `db.CommonExtras`, which seems to primarily be used to add methods for handling redirects.
3. You probably want some `statsd` counts for this feature somewhere.
4. What happens when a followed patron makes their reading log private?  Do they lose all of their followers?
5. When I was on the `/followers` and `/following` pages, it seemed odd that these options were not present in the sidebar.

Of these, I think that `#1` is a show-stopper until fixed, and `#4` seems very important and may have policy implications.

**Comment by jimchamp** [comment]
Merging this should also close #1966.

**Review by jimchamp — APPROVED** [review]
Thanks @mekarpeles!  This largely looks good to me.

One thing: While, testing, I didn't realize that these were links until the cursor changed on hover.  Patrons using mobile devices may never make such a realization.  Also, the clickable region for each is large:
image

Not a show-stopper, and maybe even a good first issue for the next iteration.

Anyway, I'm going to approve this PR, but it is blocked from merging until #9147 is merged.  Once that is done, feel free to merge this yourself, @mekarpeles.

**Pull request merged**